pdns/recursordist/rpzloader.cc

```diff
@@ -445,176 +459,252 @@ void RPZIXFRTracker(const std::vector<ComboAddress>& primaries, const boost::opt
+  }
+  catch (const PDNSException& e) {
+    SLOG(g_log << Logger::Error << "Error while applying the update received over XFR for " << zoneName << ", skipping the update: " << e.reason << endl,
+         logger->error(Logr::Error, e.reason, "Exception while applying the update received over XFR, skipping", "exception", Logging::Loggable("PDNSException")));
+  }
+  return true;
+}
+
+// As there can be multiple threads doing updates (due to config reloads), we use a multimap.
+// The value contains the actual thread id that owns the struct.
+
+static LockGuarded<std::multimap<DNSName, RPZWaiter&>> condVars;
+
+// Notify all threads trakcing the RPZ name
+bool notifyRPZTracker(const DNSName& name)
+{
+  auto lock = condVars.lock();
+  auto [start, end] = lock->equal_range(name);
+  if (start == end) {
+    // Did not find any thread tracking that RPZ name
+    return false;
+  }
+  while (start != end) {
+    start->second.stop = true;
+    start->second.condVar.notify_one();
+    ++start;
+  }
+  return true;
+}
+
+void RPZIXFRTracker(RPZTrackerParams params, uint64_t configGeneration)
+{
+  setThreadName("rec/rpzixfr");
+  bool isPreloaded = params.soaRecordContent != nullptr;
+  auto logger = g_slog->withName("rpz");
+  RPZWaiter waiter(std::this_thread::get_id());
+
+  /* we can _never_ modify this zone directly, we need to do a full copy then replace the existing zone */
+  std::shared_ptr<DNSFilterEngine::Zone> oldZone = g_luaconfs.getLocal()->dfe.getZone(params.zoneIdx);
+  if (!oldZone) {
+    SLOG(g_log << Logger::Error << "Unable to retrieve RPZ zone with index " << params.zoneIdx << " from the configuration, exiting" << endl,
+         logger->error(Logr::Error, "Unable to retrieve RPZ zone from configuration", "index", Logging::Loggable(params.zoneIdx)));
+    return;
+  }
+
+  // If oldZone failed to load its getRefresh() returns 0, protect against that
+  uint32_t refresh = std::max(params.refreshFromConf != 0 ? params.refreshFromConf : oldZone->getRefresh(), 10U);
+  DNSName zoneName = oldZone->getDomain();
+  std::string polName = !oldZone->getName().empty() ? oldZone->getName() : zoneName.toStringNoDot();
+
+  // Now that we know the name, set it in the logger
+  logger = logger->withValues("zone", Logging::Loggable(zoneName));
+
+  {
+    auto lock = condVars.lock();
+    lock->emplace(zoneName, waiter);
+  }
+  preloadRPZFIle(params, zoneName, oldZone, refresh, polName, waiter, logger);
+
+  bool skipRefreshDelay = isPreloaded;
+
+  while (RPZTrackerIteration(params, zoneName, oldZone, refresh, polName, skipRefreshDelay, configGeneration, waiter, logger)) {
+    // empty
+  }
+
+  // Zap our (and only our) RPZWaiter struct out of the multimap
+  auto lock = condVars.lock();
+  auto [start, end] = lock->equal_range(zoneName);
+  while (start != end) {
+    if (start->second.id == std::this_thread::get_id()) {
+      start = lock->erase(start);
```

Comment:
```suggestion
      start = lock->erase(start);
      break;
```

Reply:
There can be multiple thread handling the same rpz (e.g. old ones that are still sleeping and will conclude it's time to stop when woken). Notifying all of them has the consequence that the old thread cleans itself up immediately and that a newer thread for the same zone gets refreshed as well.

Reply:
But there is only one entry with our thread ID, right?

Reply:
Right. But the key into the `multimap` is a `DNSName`

Reply:
I'm sorry to be so thick but I I'm afraid I still don't get it :-) This loop is iterating over the multimap to remove the entry corresponding to our thread, so I was assuming that once we have done that there is no need to iterate over the remaining entries, but perhaps I'm wrong about that?

Reply:
Consider the scenario with an Lua reload. That wil create a new thread serving RPZ zone `X`. The old thread will also doing that and will not quit until it awakens. So after the reload there are two entries serving `X` in the multimap. 

When a notify comes in and we iterate over those two entries, sending then both a `notify_one`. The first thread will wake up, see its config is outdated by the generation mechanism and remove itself from the multimap and quit, the second will wake up and trigger a refresh without removing itself.

Reply:
Oops, yes, sorry about that, was looking at the wrong loop in the the editor :-(. Your suggestion is correct.